### PR TITLE
Assosciations setup

### PR DESCRIPTION
**Pull request by BenM4k**
In this update, I:
- Create a class Classroom 
- Create the has-many/belongs-to relationship between Classroom and Student. 
- Create the belongs-to side 
- Made sure that when adding a student to a classroom it also sets the classroom for the student.
- Made sure that when setting the classroom for a student it also adds it to the classroom students.

- Create a class Book
- Create a class Rental
- Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental.

For more details, click [here](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md)

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_accessor :label
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def many?(student)
```
Comment (aamirkhan2478) [inline classroom.rb]: - [ ] Please rename the method `many?` to `add_student` to make it clearer that it's adding a student to the classroom.

File: student.rb [hunk 2]
```diff
@@ -9,4 +11,8 @@ def initialize(id, age, name, classroom)
   def play_hooky
     '¯\(ツ)/¯'
   end
+
+  def belongs_to?
+    @classroom = self
+  end
```
Comment (aamirkhan2478) [inline student.rb]: - [ ] Please follow the steps below :point_down:  to create the correct method 

```suggestion
def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

- **def classroom=(classroom):**

  - This defines a method named classroom= which takes one argument, classroom.
- **@classroom = classroom:**

  - This assigns the value of the classroom argument to the instance variable @classroom.
- **classroom.students.push(self) unless classroom.students.include?(self):**

  - **This line has two parts:**
    - **classroom.students.push(self):** It adds the current instance (self) to the list of students in the provided classroom.
    - **unless classroom.students.include?(self):** This part checks whether the current instance (self) is already present in the list of students for the provided classroom. If it's not present, the push operation is performed to add it to the list.

In summary, this method is used to set the classroom for a student instance and also ensures that the student is added to the list of students for that classroom if they are not already present. It prevents adding the same student to the classroom's student list multiple times.

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
```
Comment (aamirkhan2478) [inline book.rb]: [OPTIONAL] kindly add `:rentals` accessor after `:author` because we want to access rentals in this class so kindly add it

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def many?(rental)
+    @rentals.push(rental)
+    rental.belongs_to = self
+  end
```
Comment (aamirkhan2478) [inline book.rb]: - [ ] Please follow the steps below 👇 to create the correct method

```suggestion
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

- **def add_rental(person, date):** 
  - This method allows you to add a rental for the `book`. It takes two arguments: `person` (the person renting the book) and `date` (the rental date).

- **Rental.new(date, self, person):** 
  - Inside the add_rental method, a new instance of the Rental class is created using the `Rental.new` constructor. This line establishes the association between the `book`, the `person`, and the rental `date`. The self represents the current book instance (the one on which the add_rental method is called).

File: person.rb [hunk 5]
```diff
@@ -20,6 +21,11 @@ def correct_name
     @name
   end
 
+  def many?(rental)
+    @rentals.push(student)
+    rental.belongs_to = self
+  end
```
Comment (aamirkhan2478) [inline person.rb]: - [ ] Please follow the steps below 👇 to create the correct method

```suggestion
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

- **def add_rental(person, date):** 
  - This method allows you to add a rental for the `book`. It takes two arguments: `person` (the person renting the book) and `date` (the rental date).

- **Rental.new(date, self, person):** 
  - Inside the add_rental method, a new instance of the Rental class is created using the `Rental.new` constructor. This line establishes the association between the `book`, the `person`, and the rental `date`. The self represents the current book instance (the one on which the add_rental method is called).

File: rental.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    @person = person
+  end
+
+  def belongs_to?
+    @book = self
+  end
+
+  def belongs_to_a_person?
+    @person = self
+  end
```
Comment (aamirkhan2478) [inline rental.rb]: - [ ] Please check below :point_down: my suggestion to correct your code

```suggestion
  attr_accessor :date, :person, :book

  def initialize(date, person, book)
    @date = date
    @book = book
    book.rental.push(self) unless book.rental.include?(self)
    @person = person
    person.rental.push(self) unless person.rental.include?(self)
  end
```

- **attr_accessor :date, :person, :book :**
  - This line creates getter and setter methods for the instance variables date, person, and book.

- **def initialize(date, person, book):**
  - This is the constructor (initializer) method for the Rental class. It takes three arguments: date, person, and book.

- **@date = date:**
  - This assigns the value of the date argument to the instance variable @date.

- **@book = book:**
  - This assigns the value of the book argument to the instance variable @book.

- **book.rentals.push(self) unless book.rentals.include?(self):**
  - This line adds the current Rental instance (self) to the rentals array of the associated book (book).
  - The unless book.rentals.include?(self) part ensures that the rental is not added multiple times to the book.rentals array.

- **@person = person:**
  - This assigns the value of the person argument to the instance variable @person.

- **person.rentals.push(self) unless person.rentals.include?(self):**
  - This line adds the current Rental instance (self) to the rentals array of the associated person (person).
  - Similar to the previous case, unless person.rentals.include?(self) prevents duplicate entries in the person.rentals array.

In summary, this initialize method creates a Rental instance and associates it with both a person and a book. It ensures that the rental is added to the corresponding arrays of rentals for the associated person and book, avoiding duplicates. This allows you to track which rentals are associated with each person and book.

**Review by aamirkhan2478 — CHANGES_REQUESTED** [review]
Hi @BenM4k, 



Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
✔️ Good Effort
✔️ Your design is good.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ridwanediallo — APPROVED** [review]
# STATUS: APPROVED 🟢
## Hi @BenM4k, 
Congratulations!GIF

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used GitHub flow correctly. ✔
- Well done implementing the changes requested by the previous reviewer. 🎉 🎉 🎉 ✔

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
